backend/peephole/peephole_rules.ml

```diff
@@ -72,113 +72,63 @@ let remove_useless_mov (cell : Cfg.basic Cfg.instruction DLL.cell) =
+    <op1> and <op2> are associative binary operators such
+    that either <op1> is the same as <op2>, or <op1> is the inverse of <op2>,
+    or there exists const3 such that <op1 const1> can be expressed as <op2 const3>
+    or <op2 const2> can be expressed as <op1 const3> *)
 
 let are_compatible op1 op2 imm1 imm2 :
     (Simple_operation.integer_operation * int) option =
   match
     ( (op1 : Simple_operation.integer_operation),
       (op2 : Simple_operation.integer_operation) )
   with
-  (* Folding two bitwise operations such as (AND, OR, XOR) should never produce
-     an overflow so we assert this conditon. *)
-  | Iand, Iand ->
-    assert (U.amd64_imm32_within_bounds imm1 imm2 ( land ));
-    Some (Iand, imm1 land imm2)
-  | Ior, Ior ->
-    assert (U.amd64_imm32_within_bounds imm1 imm2 ( lor ));
-    Some (Ior, imm1 lor imm2)
-  | Ixor, Ixor ->
-    assert (U.amd64_imm32_within_bounds imm1 imm2 ( lxor ));
-    Some (Ixor, imm1 lxor imm2)
+  | Iand, Iand -> U.bitwise_immediates op1 imm1 imm2 ( land )
+  | Ior, Ior -> U.bitwise_immediates op1 imm1 imm2 ( lor )
+  | Ixor, Ixor -> U.bitwise_immediates op1 imm1 imm2 ( lxor )
   (* For the following three cases we have the issue that in some situations,
      one or both immediate values could be out of bounds, but the result might
      be within bounds (e.g. imm1 = -4 and imm2 = 65, their sum being 61). This
      should not happen at all since the immediate values should always be within
      the bounds [0, Sys.int_size]. *)
-  | Ilsl, Ilsl ->
-    if Misc.no_overflow_add imm1 imm2 && imm1 + imm2 <= Sys.int_size
-    then (
-      U.bitwise_shift_assert imm1 imm2;
-      Some (Ilsl, imm1 + imm2))
-    else None
-  | Ilsr, Ilsr ->
-    if Misc.no_overflow_add imm1 imm2 && imm1 + imm2 <= Sys.int_size
-    then (
-      U.bitwise_shift_assert imm1 imm2;
-      Some (Ilsr, imm1 + imm2))
-    else None
-  | Iasr, Iasr ->
-    if Misc.no_overflow_add imm1 imm2 && imm1 + imm2 <= Sys.int_size
-    then (
-      U.bitwise_shift_assert imm1 imm2;
-      Some (Iasr, imm1 + imm2))
-    else None
-  (* for the amd64 instruction set the `ADD` `SUB` `MUL` opperations take at
-     most an imm32 as the second argument, so we need to check for overflows on
-     32-bit signed ints. *)
-  (* CR-someday gtulba-lecu: This condition is architecture specific and should
-     either live in amd64 specific code or this module should contain
-     information about the architecture target. *)
-  | Iadd, Iadd ->
-    if Misc.no_overflow_add imm1 imm2
-       && U.amd64_imm32_within_bounds imm1 imm2 ( + )
-    then Some (Iadd, imm1 + imm2)
-    else None
+  | Ilsl, Ilsl | Ilsr, Ilsr | Iasr, Iasr | Iadd, Iadd ->
+    U.add_immediates op1 imm1 imm2
   | Iadd, Isub ->
+    (* CR gyorsh: The following transformation changes the order of operations
+       on [r] and therefore might change the overflow behavior: if [r+c1]
+       overflows, but r-[c2-c1] does not overflow. *)
```

Comment:
I think this is addressed by the
z3/smtlib encoding I posted on
the pull request, but please
double check the encoding is
correct.